### PR TITLE
TASK: Routing: Throw exception for optional route parts without default

### PR DESCRIPTION
**Pull request by bwaidelich**
In order to match/resolve unambiguously for every dynamic optional route
part a corresponding default value has to be specified:

```
-
  uriPattern: 'foo/({bar})'
  defaults:
    'bar': 'without-this-it-wont-work'
```

With this change an exception is thrown when a route with optional dynamic
parts without corresponding defaults is configured.

Resolves: #683

### CODE REVIEW

**Comment by kitsunet** [comment]
Seems some test nicely exposes the new behavior and needs adaption...

**Comment by bwaidelich** [comment]
> Seems some test nicely exposes the new behavior and needs adaption...

I don't understand?

File: TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php [hunk 1]
```diff
@@ -559,8 +559,9 @@ public function routeMatchesRequestPathWithAllPartsIfUriPatternContainsTwoOption
 
     /**
      * @test
+     * @expectedException \TYPO3\Flow\Mvc\Exception\InvalidRouteSetupException
```
Comment (bwaidelich) [inline TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php]: This is new. With this adjusted test we now test the new behavior

File: TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php [hunk 2]
```diff
@@ -581,16 +582,6 @@ public function routeMatchesEmptyRequestPathIfUriPatternContainsOneOptionalDynam
     /**
      * @test
      */
-    public function routeDoesNotMatchRequestPathContainingNoneOfTheOptionalRoutePartsIfNoDefaultsAreSet()
```
Comment (bwaidelich) [inline TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php]: These don't make sense any longer because this case can't ever happen

**Comment by bwaidelich** [comment]
@hhoechtl @kitsunet *ping*

**Comment by kitsunet** [comment]
Sorry, I restarted tests, but they are still red, I am unsure about your explanation. Shouldn't the tests all run?

**Comment by bwaidelich** [comment]
> Shouldn't the tests all run?

Yes, they should. The failing travis CI is unrelated:
image

**Comment by kitsunet** [comment]
Huh, but that is just the line that says that there is a test failure. I see this:

```
1) TYPO3\Flow\Tests\Functional\Mvc\ActionControllerTest::defaultActionSpecifiedInrouteIsCalledAndResponseIsReturned

Failed asserting that two strings are equal.

--- Expected

+++ Actual

@@ @@

-'First action was called'

+'

+Uncaught Exception in Flow #1477140679: There is no default value specified for the optional route part "{@action}" of route "testb", but all dynamic optional route parts need a default.

+thrown in file /home/travis/build/neos/flow-development-distribution/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/TYPO3_Flow_Mvc_Routing_Route.php

+in line 643

+
```

**Comment by bwaidelich** [comment]
@kitsunet doh, I must be blind ;)
I fixed the test now and rebased the PR (the failing tests are now really unrelated).
PS: This PR can be squashed on merge

**Comment by albe** [comment]
Hm, the test failue still states

```
There was 1 failure:
1) TYPO3\Flow\Tests\Functional\Mvc\ViewsConfiguration\ViewsConfigurationTest::templatePathAndFilenameIsChanged
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Changed on Package Level'
+'
+Uncaught Exception in Flow #1477140679: There is no default value specified for the optional route part "{@action}" of route "viewsconfigurationa", but all dynamic optional route parts need a default.
+thrown in file /home/travis/build/neos/flow-development-distribution/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/TYPO3_Flow_Mvc_Routing_Route.php
+in line 643
```

Still related I think ;)

**Comment by bwaidelich** [comment]
uhm, this was getting embarrassing.. now for good! ;)

**Comment by albe** [comment]
No worries, merges always carry possibility of (re)introducing wrong behaviour :) Looks good now

**Pull request merged**